crates/services/gas_price_service/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+use async_trait::async_trait;
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    StateWatcher,
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use std::sync::{
+    Arc,
+    RwLock,
+};
+
+pub mod static_updater;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Failed to update gas price algorithm")]
+    UpdateFailed(String),
+}
+
+pub struct GasPriceService<A, U> {
+    next_block_algorithm: Arc<RwLock<(BlockHeight, A)>>,
+    update_algorithm: U,
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+{
+    pub fn new(starting_block_height: BlockHeight, update_algorithm: U) -> Self {
+        let algorithm = update_algorithm.start(starting_block_height);
+        let next_block_algorithm =
+            Arc::new(RwLock::new((starting_block_height, algorithm)));
+        Self {
+            next_block_algorithm,
+            update_algorithm,
+        }
+    }
+
+    pub fn next_block_algorithm(&self) -> Arc<RwLock<(BlockHeight, A)>> {
+        self.next_block_algorithm.clone()
+    }
+}
+
+#[async_trait]
+pub trait UpdateAlgorithm {
+    type Algorithm;
+
+    fn start(&self, for_block: BlockHeight) -> Self::Algorithm;
+    async fn next(&mut self, for_block: BlockHeight) -> Self::Algorithm;
+}
+
+fn next_block_height(block_height: &BlockHeight) -> BlockHeight {
+    (u32::from(*block_height) + 1).into()
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+{
+    fn update(
+        &self,
+        new_block_height: BlockHeight,
+        new_algorithm: A,
+    ) -> Result<(), Error> {
+        let mut latest_algorithm = self
+            .next_block_algorithm
+            .write()
+            .map_err(|e| Error::UpdateFailed(format!("{e:?}")))?;
+        let (block_height, algorithm) = &mut *latest_algorithm;
+        *algorithm = new_algorithm;
+        *block_height = new_block_height;
+        Ok(())
+    }
+
+    fn latest_block_height(&self) -> BlockHeight {
+        self.next_block_algorithm.read().unwrap().0
+    }
+}
+
+#[async_trait]
+impl<A, U> RunnableService for GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A> + Send,
+    A: Send + Sync,
+{
+    const NAME: &'static str = "GasPriceUpdater";
+    type SharedData = Arc<RwLock<(BlockHeight, A)>>;
```

Comment:
It would be nice if we had a separate struct maybe `struct SharedState`( or any other name=)) which contains all shared fields.

Reply:
Yeah. It would be nice, if only to have getters etc. The provider should be able to depend on the service types without breaking dependency inversion. So I can do that.